### PR TITLE
feat: add mongodb configuration overloads

### PR DESCRIPTION
**Pull request by ailtonguitar**
**What this PR does / why we need it**:
Adds new overloads to mongodb health check, it is necessary to give support for reusing mongoclient and mongoclientsetting instances from the container and take advantage of advanced configurations like tracing and metrics.

**Does this PR introduce a user-facing change?**:
No

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [ ] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### CODE REVIEW

File: src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs [hunk 1]
```diff
@@ -190,4 +190,124 @@ public static class MongoDbHealthCheckBuilderExtensions
...
+    /// <param name="failureStatus">
+    /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+    /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+    /// </param>
+    /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+    /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
+    /// <returns>The specified <paramref name="builder"/>.</returns>
+    public static IHealthChecksBuilder AddMongoDb(
+        this IHealthChecksBuilder builder,
+        Func<IServiceProvider, MongoClientSettings> mongoClientSettingsFactory,
```
Comment (eerhardt) [inline src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs:209]: How common is this scenario? (Having the MongoClientSettings come from an IServiceProvider). Is it something we could live without these overloads? Someone could easily do this themselves since MongoDbHealthCheck is public.

Comment (ailtonguitar) [reply src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs]: I can see some use cases for that, the idea here was to give more flexibility and different options to configure the healthcheck.
But I think we can live without that and I've already removed the overloads.
Please review.

Comment (ailtonguitar) [reply src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs]: Hi @eerhardt can we close this one?

**Comment by ailtonguitar** [comment]
Can we move forward with this PR? Is something missing here?

File: src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs (hunk 1 above)
Comment (eerhardt) [reply src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs]: Yep, resolved.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution.

@unaizorrilla @sungam3r @adamsitnik - any thoughts here?

File: src/HealthChecks.MongoDb/MongoDbHealthCheck.cs [hunk 2]
```diff
@@ -21,6 +21,12 @@ public MongoDbHealthCheck(string connectionString, string? databaseName = defaul
         }
     }
 
+    public MongoDbHealthCheck(IMongoClient client, string? databaseName = default)
+        : this(client.Settings, databaseName)
+    {
+        _mongoClient[_mongoClientSettings.ToString()] = client;
```
Comment (adamsitnik) [inline src/HealthChecks.MongoDb/MongoDbHealthCheck.cs:27]: We should remove this cache, but this is outside of the scope of this PR. I've created #2148 to track it.

**Review by adamsitnik — COMMENTED** [review]
Thank you for your contribution @ailtonguitar ! PTAL at my feedback

File: src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs [hunk 3]
```diff
@@ -190,4 +190,64 @@ public static class MongoDbHealthCheckBuilderExtensions
...
+            tags,
+            timeout));
+    }
+
+    /// <summary>
+    /// Add a health check for MongoDb database that list all collections from specified database on <paramref name="mongoDatabaseName"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+    /// <param name="mongoClientFactory">A factory to build MongoClientSettings to be used.</param>
+    /// <param name="mongoDatabaseName">The Database name to check.</param>
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs:228]: Agreed. I like @adamsitnik's suggestion.

Note that this was copied from the other overloads. They should be fixed as well.

File: src/HealthChecks.MongoDb/MongoDbHealthCheck.cs (hunk 2 above)
Comment (ailtonguitar) [reply src/HealthChecks.MongoDb/MongoDbHealthCheck.cs:27]: Hi @adamsitnik! Since it is not part of the scope of this PR can we close this one?

File: src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs [hunk 4]
```diff
@@ -190,4 +190,65 @@ public static class MongoDbHealthCheckBuilderExtensions
...
+            tags,
+            timeout));
+    }
+
+    /// <summary>
+    /// Add a health check for MongoDb database that list all collections from specified database on <paramref name="mongoDatabaseName"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+    /// <param name="mongoClientFactory">A factory to build MongoClient to be used.</param>
+
```
Comment (adamsitnik) [inline src/HealthChecks.MongoDb/DependencyInjection/MongoDbHealthCheckBuilderExtensions.cs:228]: this empty line is breaking the build ;)

```suggestion
```

**Review by adamsitnik — CHANGES_REQUESTED** [review]
We are almost there. Please update this file with the new APIs that you are introducing in this PR: https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/master/test/HealthChecks.MongoDb.Tests/HealthChecks.MongoDb.approved.txt

IIRC you should be able to run these particular tests locally

**Comment by ailtonguitar** [comment]
> We are almost there. Please update this file with the new APIs that you are introducing in this PR: https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/master/test/HealthChecks.MongoDb.Tests/HealthChecks.MongoDb.approved.txt
> 
> IIRC you should be able to run these particular tests locally

Hi @adamsitnik! Done, please review.

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you for your contribution @ailtonguitar !
